oeps/architectural-decisions/oep-XXXX-modular-micro-frontend-domains.rst

```diff
@@ -0,0 +1,209 @@
+
+* Inconsistency in deprecation efforts
+
+  * The gradual replacement of features often goes contrary to the Principle of Least Surprise: features are modified or removed with little to no warning between releases.
+
+Some of these are a natural consequence of abandoning server-side rendering; in particular, the authors do not suggest it will ever be possible for an operator to override *anything* on a page without forking the codebase, as was possible with Comprehensive Theming.  However, it seems all of the listed problems can at least be ameliorated - if not completely resolved - by implementing a combination of industry best-practices that have risen to maturity in the past couple of years, all of which introduce foreseeable costs that are clearly outweighed by their advantages.
+
+It is thus a good time to consider these practices - specified below - weighing both the development cost of implementation and future recurring ones against their benefits.
+
+
+*************
+Specification
+*************
+
+We propose a compromise between the monolithic codebase and full MFE independence based on Domain Driven Development: the Open edX frontend will be split into formal top-level domains, including but not limited to LMS and CMS, each of which will be implemented as a Container SPA.  The container, in turn, will render multiple partly-independent MFEs as sub-components.  The latter will be re-engineered to conform to the hosting domain's interfaces.
+
+This is to be a full re-platforming, to be undertaken as a single long-term project prior to release.  It is in opposition to the currently in-progress gradual conversion of single pages to fully independent MFEs that can be linked to from the pre-existing monolithic framework; the opposite, injecting legacy pages into a domain SPA as components, would be unjustifiable.  The difficulty of working with the legacy codebase is the main reason why it was decided to move to MFEs in the first place!
+
+On the other hand, the current gradual conversion to fully independent MFEs need not be interrupted; the domain-based re-platforming proposed here can be executed in parallel.  Why?  Converting an already fully independent MFE to a domain-hosted one is much less costly than the initial corresponding extraction from the monolith.  And conversely, it is conceivable that a particular monolithic page, after extraction directly into a domain-hosted MFE, could then be cheaply backported *the other way*, into a fully-independent one for earlier release while the re-platforming project is incomplete.
+
+As this OEP is developed, we will further investigate:
+
+#. Integration strategy
+#. Domain SPAs and their interfaces
+#. Shared dependencies
+#. MFEs as components
+
+Discovery
+=========
+
+* Are there any further top-level architectural elements to consider?
+* What is the relationship of this project to the Modular Learning effort?
```

Comment:
[curious] What is the "Modular Learning effort"?

Reply:
My gut feelling is that it realtes to this https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3470196737/Modular+Learning